Standardise interface for accessing the owner of a symbol

thomasspriggs opened the pull request:
Without the functions added by this PR, the only way to access the
owner of a symbol for static fields and java methods is through
`symbol.type.get(ID_C_class)`. When this technique is used by downstream
repositories, they are vulnerable to being broken by upstream changes to
how this information is stored in JBMC. The intention of this PR is
that the two new functions added will be the standard interface to this
symbol -> owner link. This means that future changes to how this
information is stored can be made without breaking downstream
repositories by changing these two functions.

This PR includes changes to use the two new functions. This is preparatory 
work ready for moving the link from being stored in a comment
on a symbols type, to the symbol itself, as discussed with @smowton 

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

thomasspriggs committed:
Add owning class functions as interface to the owner of a symbol

Without the functions added by this commit, the only way to access the
owner of a symbol for static fields and java methods is through
`symbol.type.get(ID_C_class)`. When this technique is used by downstream
repositories, they are vulnerable to being broken by upstream changes to
how this information is stored in JBMC. The intension of this commit is
that the two new functions added will be the standard interface to this
symbol -> owner link. This means that future changes to how this
information is stored can be made without breaking downstream
repositories by changing these two functions.

thomasspriggs committed:
Use owning class functions as interface to the owner of a symbol

This commit replaces access to `symbol.type.get(ID_C_class)`. With calls
to the `owning_class` function and access to
`symbol.type.set(ID_C_class, class_id)` with calls to
`set_owning_class`. This allows for changes to how the owner of a symbol
is stored by changing these two functions only, rather than by changing
all the places where `ID_C_class` was previously accessed directly. This
is preparatory work for moving this link from being stored in a comment
on a symbols type, to the symbol itself.

tautschnig reviewed (dismissed):
What is meant by the "owner of a symbol?" For a given symbol `x`, Is it the class that `x` is a member of? But then those aren't separate symbols, they would just be members of a struct type. So I'm confused.

Edit: the only owner of a symbol that we have is the symbol table.

thk123 commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -419,6 +419,8 @@ void java_bytecode_convert_methodt::convert(
     get_method_identifier(class_symbol.name, m);
 
   method_id=method_identifier;
+  set_owning_class(
```
:pick: Lower down (line 537) you have a writeable ref for this symbol already which is used to update the type. It might be nicer to keep the mutation in one place and do this call down there

thomasspriggs replied:
I tried putting it lower down and things didn't work properly. I guess it must be read back somewhere between the two locations.

thk123 commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -464,3 +464,14 @@ symbolt &fresh_java_symbol(
+optionalt<irep_idt> owning_class(const symbolt &symbol)
+{
+  const irep_idt &class_id = symbol.type.get(ID_C_class);
+  return class_id.empty() ? optionalt<irep_idt>{} : class_id;
```
:pick: can `optionalt<irep_idt>{}` just be `{}`?

thomasspriggs replied:
Nope, because of the usage of the ternary. I'd have to refactor this into an `if` instead in order to be able to use just `{}`.

thk123 reviewed (commented):
Code changes look good, but I think to make this robust, it needs a unit test that actually goes through the java parser and checks the properties of a real static symbol. At a minimum java class with a static field and a static method.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: bd2bacd2e13387adb0658ea0c9ae7d83b4311540).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106300980
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

thomasspriggs committed:
Rename `owning_class` to `declaring_class`

thomasspriggs committed:
Fix missing declaring class of opaque methods

thomasspriggs committed:
Add unit tests covering method parsing setting `declaring_class`

thomasspriggs committed:
Test that parsed static fields have correct `declaring_class`

thomasspriggs commented:
I have
- Renamed `owning_class` to `declaring_class`, as discussed with @tautschnig 
- Added unit tests covering the java parsing setting `declaring_class` as requested by @thk123

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 65e3d31a79aa9c439b7bad473509a93d48092783).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106341108
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
